tests/Unit/CallbacksRegistrarTest.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+namespace LaraCrafts\GeoRoutes\Tests\Unit;
+
+use Closure;
+use LaraCrafts\GeoRoutes\CallbacksRegistrar;
+use LaraCrafts\GeoRoutes\Tests\TestCase;
+use Mockery;
+use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
+use ReflectionClass;
+
+class CallbacksRegistrarTest extends TestCase
+{
+    use MockeryPHPUnitIntegration;
+
+    /** @var \Mockery\MockInterface */
+    protected $location;
+
+    /** @var \Illuminate\Routing\Route */
+    protected $route;
+
+    /** @var \Illuminate\Routing\Router */
+    protected $router;
+
+    public function setUp(): void
+    {
+        parent::setUp();
+    }
+
+    public function tearDown(): void
+    {
+        Mockery::close();
+    }
+
+    public function testIfConfigCallbacksAreLoaded()
+    {
+        $callbacks = [
+            'foo' => '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::foo',
+            'bar' => '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::bar',
+        ];
+
+        $registrar = Mockery::mock(CallbacksRegistrar::class)->makePartial();
+
+        $registrar->shouldReceive('addCallback')
+                ->ordered()
+                ->once()
+                ->with('foo', '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::foo');
+
+        $registrar->shouldReceive('addCallback')
+                ->ordered()
+                ->once()
+                ->with('bar', '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::bar');
+
+        $registrar->__construct($callbacks);
```

Comment:
Why would you do this?

Reply:
Yes, there is a way to do that. I have done a small research and found out that it is, most of the times, considered a bad design when calling methods from the *constructor*, especially when it makes testing more complicated (in our case it does, because when creating a mock we have to ignore the constructor then explicitly calling it).

So i thought we could change the visibility of the [`loadProxies`](https://github.com/LaraCrafts/laravel-geo-routes/pull/36/files/#diff-5d38219739bdd7ea4a7e4beb460004f6R128) method to `public`, making it depend on a `callbacks` property that has to be set in the *constructor* and calling it when registering our binding, which will result in the following code:
```php
        $this->app->singleton('georoutes.callbacks', function () {
            $registrar = new CallbacksRegistrar(config('geo-routes.routes.callbacks'));

            $registrar->loadProxies();

            return $registrar;
        });
```
In this way our *constructor* will look like the following:
```php
    public function __construct(array $callbacks)
    {
        $this->proxies = [];
        $this->callbacks = $callbacks;
    }
```
And this is the final test method code:
```php
    public function testIfConfigCallbacksAreLoaded()
    {
        $callbacks = [
            'foo' => '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::foo',
            'bar' => '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::bar',
        ];

        $registrar = Mockery::mock(CallbacksRegistrar::class, [$callbacks])->makePartial();

        $callbacksProperty = $this->getProperty($registrar, 'callbacks');
        # Assert that the callbacks property is being properly set in the constructor
        $this->assertEquals($callbacks, $callbacksProperty->getValue($registrar));

        $registrar->shouldReceive('addCallback')
                ->ordered()
                ->once()
                ->with('foo', '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::foo');

        $registrar->shouldReceive('addCallback')
                ->ordered()
                ->once()
                ->with('bar', '\LaraCrafts\GeoRoutes\Tests\Mocks\Callbacks::bar');

        $registrar->loadProxies();
    }
```

I honestly think this code looks much better than the current.

Reply:
Have you tried [Generated Partial Test Doubles](http://docs.mockery.io/en/latest/reference/partial_mocks.html#generated-partial-test-doubles). According to the docks that will call the original constructor so maybe you can call `loadProxies` from there.

Reply:
It is not a matter of whether we can call the *constructor* while creating the *mock* or not, actually we can directly call the *constructor* by passing its arguments (e.g. `Mockery::mock(CallbacksRegistrar::class, [$callbacks])`), the issue here is whenever the `__construct` method is called, the `addCallback` method is called as well while there are no expectations to that. Also mocking the `loadProxies` method will not do us any good, because the actual code will never be executed and therefore our `addCallback` method assertions will fail. I tried using a *`spy`* too but for some reason it is not working either. I am sure in a way or another we can test the current code without having to explicitly call the *constructor*, however i think the new design makes things way better than they are right now, i mean don't you think so? I would love to hear your opinion concerning this new design i am proposing.

Reply:
It is a viable solution. My opinion is if we can avoid a "you must call this method to make it work" scenario then avoid it.

Reply:
Honestly i don't see it as a "you must call this method to make it work" scenario, because the developer does not have to call the `loadProxies` method at all, the first time the binding is resolved the proxies will be automatically loaded after the instance is created. Even in our [*GeoRoute*](https://github.com/LaraCrafts/laravel-geo-routes/blob/callbacks-registrar/src/GeoRoute.php) class we are only using the Facade ([here](https://github.com/LaraCrafts/laravel-geo-routes/blob/b5211c53e22d51e40a1d80a986b301284b87d047/src/GeoRoute.php#L80)), so the only place where we used the `loadProxies` method is when [defining the singleton](https://github.com/LaraCrafts/laravel-geo-routes/blob/b5211c53e22d51e40a1d80a986b301284b87d047/src/GeoRoutesServiceProvider.php#L37).
In addition to that, i have tried several ways, including using the PHPUnit mock builder, prophecy spies and more complicated stuff like generating a mock using the PHPUnit mock builder and "prophesize-ing" it, with no success. So at the moment, i think i should take back this part
> I am sure in a way or another we can test the current code without having to explicitly call the *constructor*

because honestly after what i have tried, i now think it is impossible to perform this test without explicitly calling the *constructor* at least with the libraries at hand.

Reply:
In that case, are you going to make `loadProxies` public and rely on that call?

Reply:
Yes, however this is of course still open for suggestions and further changes if necessary.

Reply:
We might also have the `loadProxies` method accept a `$callbacks` argument, and make it another way to load an array of callbacks. What do you think?

Reply:
One thing, that might simplify everything. Do we really need to test if `addCallback` is called or is it enough to to check if `callbacks` returns the expected result?

Reply:
Yes it is crucial for our code coverage, after all we are aiming for a better code coverage percentage. And i honestly think it is a bad design whenever we have to alter our tests for the sake of a malformed class.